### ISSUE TITLE
Create large parallel NC mesh "on the fly" using material function

### ISSUE THREAD

**Issue by Maatinski** [mfem/mfem#3247]
Hi all,
This post is related to https://github.com/mfem/mfem/issues/2759 (I initially posted this there, but this might deserve its own post, as I think this might be relevant to others).

I finally got a chance to look into implementing @jakubcerveny 's proposed approach (in https://github.com/mfem/mfem/issues/2759) to generate large meshes (especially nonconforming) "on the fly" using a material function, mostly to reduce the memory load while the mesh is partitioned. This is similar to the `shaper` miniapp.
The approach is loading a material function, essentially a voxel map on a uniform 3D Cartesian grid (one attribute per Cartesian point, i.e., not a continuous function, with 0 = background) directly in MFEM and creating a coarse mesh (say 4x4x4), then partitioning the coarse mesh, and then performing the NC refinement iteratively (similarly to the `shaper ` miniapp). This seems to work quite well (at the moment in serial, but I assume partitioning the coarse mesh, and then refining in parallel + re-balancing should work just as fine).
The only issue I noticed: if I start with a coarse mesh and then perform iterative refinement, I will end up with mesh cells at the boundary of the domain that have an undefined attribute. Here is an example:
![image](https://user-images.githubusercontent.com/63864061/193859600-c0bc6ba2-3945-4721-86b2-b4d4f20ba053.png)

![image](https://user-images.githubusercontent.com/63864061/193859937-b67587a0-3c15-4aa2-a14f-6c6e307b29fe.png)

The blue elements on the boundary have an attribute of zero. I assume I could simply set those to have zero conductivity (since I am solving a low-frequency quasi-static eddy-current problem), but I am not sure if this will distort the solution. In any case it will add unneeded complexity.
Is there a way to remove these elements from the final mesh?

I will probably run into some other issues as I am moving forward and will post some updates here.

Thanks a lot and warm regards,
Mathias

**Comment by mlstowell** [comment]
Hello, Mathias (@Maatinski)

I haven't dug into the non-conforming mesh code in quite some time but I think this is not currently possible. I believe the assumption is that the coarse mesh defines the full domain. In your use case I can see that the domain boundary would be updated as the mesh is refined which would leave portions of the original coarse mesh outside the domain.

It might be possible to tag certain child elements as being not part of the domain. The more difficult part might be updating the list of boundary elements because the location of the boundary would be altered. I suspect this moving boundary issue would prevent a quick and easy change to the `NCMesh` class to support this use case.

There are probably a number of workarounds you could try. Special materials, as you suggest, would probably be the most common approach. Another possibility is to set the degrees of freedom in these blue elements to prescribed values using a Dirichlet BC. The trick, as you know, is finding something that imposes a suitable boundary condition.

Best wishes,
Mark

**Comment by Maatinski** [comment]
Hi @mlstowell 
Thanks a lot for the clarifications. It sounds like this problem has a solution; I will try your suggestion, as soon as I have a "solvable" mesh.

I did run into a few other problems, when changing to parallel meshes. In principle everything seems to work fine (the code works well for one and two processes), but at >2 processes the attributes get messed up. I am attaching an example of my code (I tried to keep this example minimal), including a generation of an example Cartesian "voxel function"

**Serial (also works for two processes)**
![image](https://user-images.githubusercontent.com/63864061/195186964-934c6c68-f494-4f00-ad89-b23ee7591014.png)


**More than two processes**
![image](https://user-images.githubusercontent.com/63864061/195186846-9eaeb017-270e-4bcb-a822-8514b5adb3fa.png)

The odd thing is that I am checking for elements with zero attribute at the very end after the refinement, but GLVIS still indicates that some elements have zero attributes.

Another weird effect happens, when I run `pmesh->Rebalance();` on the mesh after NC refinement, which somehow reset all attributes that were changed during the refinement process.
![image](https://user-images.githubusercontent.com/63864061/195186498-6288f49c-0618-46a8-9bcb-0b9ed5e2cbca.png)


**Code snippet**
```cpp
// MQS - Magnetic Quasistatic solver for hexahedral body models
// Needs at least the following inputs:
// mesh_input  - string pointing to the body model mesh (can be .mesh, .mesh.gz., or .bin file)
// coil_input  - string pointing to the coil mesh (can be .mesh, or .txt file with multiple coil mesh paths)
//               Coil meshes are represented in mfem format using piecewise straight current carrying segments
//               These are used for Biot-Savart calculations of the magnetic vector potential driving the E-fields
// sigma_input - string pointing to a .txt file with a list of conductivity values for each tissue class (i.e., mesh attribute)

#include "mfem.hpp"
#include <fstream>
#include <iostream>
#include <string>
#include <unistd.h>
#include <omp.h>

using namespace std; // @suppress("Symbol is not resolved")
using namespace mfem;

// mesh variables
static Mesh *mesh;
static ParMesh *pmesh;
static Array<double> vertices;
static Array<int> elements;
static Array<int> attributes;
static Array<int> boundary_indices;
static Array<int> boundary_attributes;

// physical (dielectric) properties
Array<double> sigma;

int get_voxel_value(Array<int> &vox, Vector &r, Vector &r_min, Vector &dr, Array<int> &nr)
{
    // get the linear indes
    double x_idx = ( r(0) - r_min(0) ) / dr(0);
    double y_idx = ( r(1) - r_min(1) ) / dr(1);
    double z_idx = ( r(2) - r_min(2) ) / dr(2);
    double x_idx_rnd = round( x_idx );
    double y_idx_rnd = round( y_idx );
    double z_idx_rnd = round( z_idx );

    
    // check if query point is invalid (out of bound)
    if ( abs( x_idx - x_idx_rnd ) > 1e-6 or abs( y_idx - y_idx_rnd ) > 1e-6 or abs( z_idx - z_idx_rnd ) > 1e-6 ) 
    { 
        std::cout << std::setprecision(4) << std::fixed;
        std::cout << " " << std::endl;
	    std::cout << " ========================== " << std::endl;
        std::cout << "    Query point:  [ " << showpoint << setw(8) << r(0) << ", " << r(1) << ", " << r(2) << " ] " << std::endl;
        MFEM_ABORT("Invalid query point"); 
    }

    int vox_value;
    if ( x_idx_rnd < 0 or x_idx_rnd > nr[0] - 1 or
         y_idx_rnd < 0 or y_idx_rnd > nr[1] - 1 or
         z_idx_rnd < 0 or z_idx_rnd > nr[2] - 1 )
    {
        vox_value = 1;
    }
    else
    {
        vox_value = vox[ x_idx + y_idx * nr[0] + z_idx * nr[0] * nr[1] ];
    }
    
    // std::cout << "          Index:  [ " << showpoint << setw(8) << x_idx << ", " << y_idx << ", " << z_idx << " ] " << std::endl;    
    // std::cout << "    Voxel value:    " << showpoint << setw(2) << vox_value << std::endl;   

    return vox_value;
}

int main(int argc, char *argv[])
{

    // 1. Initialize MPI.
    int num_procs, myid;
    MPI_Init(&argc, &argv);
    MPI_Comm_size(MPI_COMM_WORLD, &num_procs);
    MPI_Comm_rank(MPI_COMM_WORLD, &myid);

    // options for mesh generator
    int NC_max_size_factor = 4;     // times the smallest voxel size
    int NC_min_size_factor = 1;     // times the smallest voxel size, keep at 1

    //  S T E P   1
    // ----------------- 
    // Create dummy voxel model (3 spheres)

    Vector dr_vox(3); dr_vox = 0.0005; // 0.5 mm min. resolution
    
    // extent of the voxel model
    Vector rr_vox_min(3); rr_vox_min = -0.02;
    Vector rr_vox_max(3); rr_vox_max = +0.02;

    // create the grid and the voxel model
    static Array<int> nr_vox(3);
    nr_vox[0] = round((rr_vox_max[0] - rr_vox_min[0])/dr_vox[0]) + 1;
    nr_vox[1] = round((rr_vox_max[1] - rr_vox_min[1])/dr_vox[1]) + 1;
    nr_vox[2] = round((rr_vox_max[2] - rr_vox_min[2])/dr_vox[2]) + 1;

    Vector xx_vox(nr_vox[0]);
    Vector yy_vox(nr_vox[1]);
    Vector zz_vox(nr_vox[2]);

    int nvox = nr_vox[0] * nr_vox[1] * nr_vox[2];
    int lin_idx = 0;
    static Array<int> voxels(nvox);
    voxels = 1;

    // three sphere with radius and center-point
    Vector center_1(3);
    center_1[0] = 0.0;
    center_1[1] = 0.0;
    center_1[2] = 0.0;
    double radius_1 = 0.02;

    Vector center_2(3);
    center_2[0] = +0.005;
    center_2[1] = +0.005;
    center_2[2] = 0.0;
    double radius_2 = 0.012;

    Vector center_3(3);
    center_3[0] = -0.008;
    center_3[1] = -0.008;
    center_3[2] = 0.0;
    double radius_3 = 0.006;

    for ( int ix = 0; ix < nr_vox[0]; ++ix ) { 
        xx_vox[ix] = rr_vox_min[0] + ix * dr_vox[0]; 
        for ( int iy = 0; iy < nr_vox[1]; ++iy ) { 
            yy_vox[iy] = rr_vox_min[1] + iy * dr_vox[1]; 
            for ( int iz = 0; iz < nr_vox[2]; ++iz ) { 
                zz_vox[iz] = rr_vox_min[2] + iz * dr_vox[2]; 
                lin_idx = ix + iy * nr_vox[0] + iz * nr_vox[0] * nr_vox[1]; 
                // index 1 reserved for background voxels resulting from the NC meshing
                if ( sqrt( (xx_vox[ix] - center_1[0])*(xx_vox[ix] - center_1[0]) + (yy_vox[iy] - center_1[1])*(yy_vox[iy] - center_1[1]) + (zz_vox[iz] - center_1[2])*(zz_vox[iz] - center_1[2]) ) < radius_1 ) {
                    voxels[lin_idx] = 2;        
                }
                if ( sqrt( (xx_vox[ix] - center_2[0])*(xx_vox[ix] - center_2[0]) + (yy_vox[iy] - center_2[1])*(yy_vox[iy] - center_2[1]) + (zz_vox[iz] - center_2[2])*(zz_vox[iz] - center_2[2]) ) < radius_2 ) {
                    voxels[lin_idx] = 3;        
                }
                if ( sqrt( (xx_vox[ix] - center_3[0])*(xx_vox[ix] - center_3[0]) + (yy_vox[iy] - center_3[1])*(yy_vox[iy] - center_3[1]) + (zz_vox[iz] - center_3[2])*(zz_vox[iz] - center_3[2]) ) < radius_3 ) {
                    voxels[lin_idx] = 4;        
                }
            }
        }
    }

    // attributes
    int vox_min_attr = voxels.Min();
    int vox_max_attr = voxels.Max();

	sigma.SetSize(vox_max_attr);
    sigma = 0.0;

	if (myid == 0)
	{
		cout << "                              " << endl;
		cout << "       MESH PREPARATION       " << endl;
		cout << " ============================ " << endl;
	}

	// some data we may need later
	int num_vert;
	int num_elem;
	int num_bndr;
	int num_attr;
	int dim = 3;

    // start timer
    double time_meshloading  = 0.0;
    tic_toc.Clear();
	tic_toc.Start();
    
    if (myid == 0) 
    {
        std::cout << std::setprecision(4) << std::fixed;
        std::cout << "                   X grid:  [ " << showpoint << setw(8) << rr_vox_min[0] << ", " << rr_vox_max[0] << " ]  dx = " << dr_vox[0] << "   N = " << nr_vox[0] << std::endl;
        std::cout << "                   Y grid:  [ " << showpoint << setw(8) << rr_vox_min[1] << ", " << rr_vox_max[1] << " ]  dy = " << dr_vox[1] << "   N = " << nr_vox[1] << std::endl;
        std::cout << "                   Z grid:  [ " << showpoint << setw(8) << rr_vox_min[2] << ", " << rr_vox_max[2] << " ]  dz = " << dr_vox[2] << "   N = " << nr_vox[2] << std::endl;
        std::cout << "               Attributes:  [ " << vox_min_attr << " ... " << vox_max_attr << " ] " << std::endl;
        std::cout << " " << endl;
    }

    //  S T E P   2
    // ----------------- 
    // Create an initial hexahedral mesh

    if (myid == 0) { cout << "   -- Creating initial mesh " << endl; }

    // get the mesh bounds and determine min/max mesh cell sizes
    Vector min_elem_size(3);
    min_elem_size[0] = NC_min_size_factor * dr_vox[0];
    min_elem_size[1] = NC_min_size_factor * dr_vox[1];
    min_elem_size[2] = NC_min_size_factor * dr_vox[2];

    Vector max_elem_size(3);
    max_elem_size[0] = NC_max_size_factor * dr_vox[0];
    max_elem_size[1] = NC_max_size_factor * dr_vox[1];
    max_elem_size[2] = NC_max_size_factor * dr_vox[2];

    Vector rr_mesh_min(3);
    Vector rr_mesh_max(3); 
    rr_mesh_min[0] = rr_vox_min[0] - dr_vox[0]/2;
    rr_mesh_min[1] = rr_vox_min[1] - dr_vox[1]/2;
    rr_mesh_min[2] = rr_vox_min[2] - dr_vox[2]/2;

    rr_mesh_max[0] = rr_mesh_min[0] + ceil((rr_vox_max[0] - rr_vox_min[0] + dr_vox[0])/max_elem_size[0])*max_elem_size[0];
    rr_mesh_max[1] = rr_mesh_min[1] + ceil((rr_vox_max[1] - rr_vox_min[1] + dr_vox[1])/max_elem_size[1])*max_elem_size[1];
    rr_mesh_max[2] = rr_mesh_min[2] + ceil((rr_vox_max[2] - rr_vox_min[2] + dr_vox[2])/max_elem_size[2])*max_elem_size[2];

    if (myid == 0) {
        std::cout << std::setprecision(4) << std::fixed;
        std::cout << "             Cell sizes X:  [ " << showpoint << setw(8) << min_elem_size[0] << ", " << max_elem_size[0] << " ]   " << std::endl;
        std::cout << "             Cell sizes Y:  [ " << showpoint << setw(8) << min_elem_size[1] << ", " << max_elem_size[1] << " ]   " << std::endl;
        std::cout << "             Cell sizes Z:  [ " << showpoint << setw(8) << min_elem_size[2] << ", " << max_elem_size[2] << " ]   " << std::endl;
        std::cout << "            Mesh bounds X:  [ " << showpoint << setw(8) << rr_mesh_min[0] << ", " << rr_mesh_max[0] << " ]   " << std::endl;
        std::cout << "            Mesh bounds Y:  [ " << showpoint << setw(8) << rr_mesh_min[1] << ", " << rr_mesh_max[1] << " ]   " << std::endl;
        std::cout << "            Mesh bounds Z:  [ " << showpoint << setw(8) << rr_mesh_min[2] << ", " << rr_mesh_max[2] << " ]   " << std::endl;
    }

    // get max size needed to store the mesh
    int num_elemx = round(( rr_mesh_max[0] - rr_mesh_min[0] ) / max_elem_size[0]);
    int num_elemy = round(( rr_mesh_max[1] - rr_mesh_min[1] ) / max_elem_size[1]);
    int num_elemz = round(( rr_mesh_max[2] - rr_mesh_min[2] ) / max_elem_size[2]);

    int num_vertx = num_elemx + 1;
    int num_verty = num_elemy + 1;
    int num_vertz = num_elemz + 1;
    num_elem = num_elemx * num_elemy * num_elemz;
    num_vert = num_vertx * num_verty * num_vertz;

    // create vertex grid
    Vector xx_vert(num_vertx);
    Vector yy_vert(num_verty);
    Vector zz_vert(num_vertz);
    for ( int ix = 0; ix < num_vertx; ++ix ) { xx_vert[ix] = rr_mesh_min[0] + ix * max_elem_size[0]; }
    for ( int iy = 0; iy < num_verty; ++iy ) { yy_vert[iy] = rr_mesh_min[1] + iy * max_elem_size[1]; }
    for ( int iz = 0; iz < num_vertz; ++iz ) { zz_vert[iz] = rr_mesh_min[2] + iz * max_elem_size[2]; }

    // create vertex list
    int elem_lin_idx = 0;
    int vert_lin_idx = 0;
    int cur_attr = 0;
    int ecnt = -1;
    bool create_elem = false;
    Vector r_cur(3);

    // iterate over all Cartesian points within the current "candidate" mesh element;
    // if any of the attributes is >1, create that cell
    vertices.SetSize(dim*num_vert); 
    elements.SetSize(8*num_elem);  
    attributes.SetSize(num_elem);       
    for ( int ix = 0; ix < num_vertx; ++ix ){
        for ( int iy = 0; iy < num_verty; ++iy ){
            for ( int iz = 0; iz < num_vertz; ++iz ){
                // create entry in vertex list
                vert_lin_idx = ix + iy * num_vertx + iz * num_vertx * num_verty;
                vertices[dim*vert_lin_idx] = xx_vert[ix];   
                vertices[dim*vert_lin_idx+1] = yy_vert[iy];  
                vertices[dim*vert_lin_idx+2] = zz_vert[iz];  

                if ( ix < num_elemx and iy < num_elemy and iz < num_elemz )
                {
                    create_elem = false;         
                    // determine if the current hex mesh element 
                    // contains any voxels with >1 attribute    
                    for ( int vx = 0; vx < NC_max_size_factor; ++vx ){
                        for ( int vy = 0; vy < NC_max_size_factor; ++vy ){
                            for ( int vz = 0; vz < NC_max_size_factor; ++vz ){
                                // sample the current attribute           
                                r_cur[0] = xx_vert[ix] + (vx+0.5) * dr_vox[0];
                                r_cur[1] = yy_vert[iy] + (vy+0.5) * dr_vox[1];
                                r_cur[2] = zz_vert[iz] + (vz+0.5) * dr_vox[2];
                                cur_attr = get_voxel_value(voxels, r_cur, rr_vox_min, dr_vox, nr_vox);
                        
                                if ( cur_attr > 1 ) { create_elem = true; break; }
                            }
                            if ( create_elem ) { break; }
                        }
                        if ( create_elem ) { break; }
                    } 

                    // don't go further if no element creation is requested
                    if ( create_elem ) 
                    {
                        // create the element based on the current indexing
                        ecnt = ecnt + 1;

                        // TODO: Can be more efficient
                        elements[8*ecnt] = ix + iy * num_vertx + iz * num_vertx * num_verty;
                        elements[8*ecnt+1] = ix + 1 + iy * num_vertx + iz * num_vertx * num_verty;
                        elements[8*ecnt+2] = ix + 1 + (iy+1) * num_vertx + iz * num_vertx * num_verty;
                        elements[8*ecnt+3] = ix + (iy+1) * num_vertx + iz * num_vertx * num_verty;
                        elements[8*ecnt+4] = ix + iy * num_vertx + (iz+1) * num_vertx * num_verty;
                        elements[8*ecnt+5] = ix + 1 + iy * num_vertx + (iz+1) * num_vertx * num_verty;
                        elements[8*ecnt+6] = ix + 1 + (iy+1) * num_vertx + (iz+1) * num_vertx * num_verty;
                        elements[8*ecnt+7] = ix + (iy+1) * num_vertx + (iz+1) * num_vertx * num_verty;
                        
                        // set atttribute
                        attributes[ecnt] = cur_attr;
                    }
                }
            }      
        }       
    }

    num_elem = ecnt+1;
    num_bndr = 0;
    
    // remove empty space in elements
    elements.SetSize(8*num_elem);  
    attributes.SetSize(num_elem);  
    boundary_indices.SetSize(0);
    boundary_attributes.SetSize(0);

    mesh = new Mesh(vertices, num_vert,
            elements, Geometry::Type::CUBE,
            attributes, num_elem,
            boundary_indices, Geometry::Type::SQUARE,
            boundary_attributes, num_bndr,
            dim, -1 );
    num_attr = mesh->attributes.Max();   
    num_bndr = mesh->GetNBE();

    if (myid == 0) {
        std::cout << "      -------------------------------- " << std::endl;
        std::cout << "       Number of elements:   " << num_elem << std::endl;
        std::cout << "          Boundary facets:   " << num_bndr << std::endl;
        std::cout << " " << std::endl;  
    }

    // important to wait for all processes here
    MPI_Barrier( MPI_COMM_WORLD );

    //  S T E P   3
    // ----------------- 
    // Reorder mesh elements using SFC approach, then perform mesh partitioning

    if (myid == 0) { std::cout << "   -- Removing unused vertices " << std::endl; }
    mesh->RemoveUnusedVertices();

    if (myid == 0) { std::cout << "   -- Reordering elements using SFC " << std::endl; }
    Array<int> element_ordering;
    mesh->GetHilbertElementOrdering(element_ordering);
    mesh->ReorderElements(element_ordering);
    
    mesh->EnsureNCMesh();
    mesh->Finalize();

    if (myid == 0) { std::cout << "   -- Partitioning mesh " << std::endl; }
    int part_method = 1;
    int *partitioning = mesh->GeneratePartitioning(num_procs, part_method);

    pmesh = new ParMesh(MPI_COMM_WORLD, *mesh, partitioning );
    delete mesh;

    cout << "               MPI Rank " << myid << ":    Number of elements = " << pmesh->GetNE() << endl;
    
    //  S T E P   4
    // ----------------- 
    // Iterative refinement

    MPI_Barrier( MPI_COMM_WORLD );
    if (myid == 0) { cout << "   -- Processing/refining mesh " << endl; }

    // Shaping loop
    int nclimit = 1;
    int iter = 0;
    bool refine_any = true;
    while ( refine_any and iter < 5 )
    {
        // assume no refinement
        iter = iter + 1;
        refine_any = false;	

        // output only for serial case
        if ( myid == 0 and num_procs == 1 ) { 
            // detailed output
            std::cout << "   Iter = "<< iter <<": Num elem. = " << setw(6) << pmesh->GetNE() << ":  ";
        }
        else if ( myid == 0 and num_procs > 1 ) { 
            // only header output
            std::cout << "      Iter = "<< iter <<": " << std::endl;
        }
        
        
        Array<Refinement> refs;

        Vector cur_size_factor_dbl(3);
        Array<int> cur_size_factor_int(3);
        const double *vert_1;
        const double *vert_2;
        const Element *cell;
        Array<int> vert_indices;
        bool refine_cur = false;
        int num_refine = 0;

        int cur_attr_1 = 0;
        int cur_attr_2 = 0;

        // iterate over all mesh elements; get the size of the current cell 
        // relative to the resolution of the Cartesian grid
        // (i.e., check if the current mesh element is of the smallest size requested)
        for (int mc = 0; mc < pmesh->GetNE(); mc++)
        {
            cell = pmesh->GetElement(mc);
            cell->GetVertices(vert_indices);

            vert_1 = pmesh->GetVertex(vert_indices[0]);
            vert_2 = pmesh->GetVertex(vert_indices[6]);      

            cur_size_factor_dbl[0] = (vert_2[0] - vert_1[0])/min_elem_size[0];
            cur_size_factor_dbl[1] = (vert_2[1] - vert_1[1])/min_elem_size[1];
            cur_size_factor_dbl[2] = (vert_2[2] - vert_1[2])/min_elem_size[2];

            if ( cur_size_factor_dbl[0] > 0.75 ) { cur_size_factor_int[0] = round( cur_size_factor_dbl[0] ); }
            else { cur_size_factor_int[0] = 0; }
            if ( cur_size_factor_dbl[1] > 0.75 ) { cur_size_factor_int[1] = round( cur_size_factor_dbl[1] ); }
            else { cur_size_factor_int[1] = 0; }
            if ( cur_size_factor_dbl[2] > 0.75 ) { cur_size_factor_int[2] = round( cur_size_factor_dbl[2] ); }
            else { cur_size_factor_int[2] = 0; }

            // iterate over all sub-points to determine whether point needs refinement
            cur_attr_1 = -1;
            cur_attr_2 = -1;
            refine_cur = false;
            int runcnt = 0;
            for ( int vx = 0; vx < cur_size_factor_int[0] and !refine_cur; ++vx ){
                for ( int vy = 0; vy < cur_size_factor_int[1] and !refine_cur; ++vy ){
                    for ( int vz = 0; vz < cur_size_factor_int[2] and !refine_cur; ++vz ){
                        runcnt = runcnt + 1;
                        // sample the current attribute           
                        r_cur[0] = vert_1[0] + (vx+0.5) * dr_vox[0];
                        r_cur[1] = vert_1[1] + (vy+0.5) * dr_vox[1];
                        r_cur[2] = vert_1[2] + (vz+0.5) * dr_vox[2];
                        if ( cur_attr_1 == -1 )  
                        { 
                            cur_attr_1 = get_voxel_value(voxels, r_cur, rr_vox_min, dr_vox, nr_vox);
                            pmesh->SetAttribute(mc, cur_attr_1);
                        }
                        else
                        {
                            cur_attr_2 = get_voxel_value(voxels, r_cur, rr_vox_min, dr_vox, nr_vox);
                        }

                        if ( cur_attr_2 > -1 and ( cur_attr_1 != cur_attr_2 ) )
                        {
                            // this means that different materials were found in the current mesh element
                            // and the cell needs to be refined
                            refine_cur = true;
                        } 
                    }
                }
            } 

            // Mark the element for refinement
            if (refine_cur)
            {
                // if at least one mesh element is refined we need to go into another iteration
                // to define correct attributes for the newly created elements
                refine_any = true;
                int type = 7;
                num_refine = num_refine + 1;
                refs.Append(Refinement(mc, type));
            }
        }
        pmesh->SetAttributes();

        // Perform refinement
        if ( myid == 0 and num_procs == 1 ) { 
            std::cout << setw(6) << refs.Size() << " marked for refinement ";
        }
        if ( refine_any ) {
            pmesh->GeneralRefinement(refs, -1, nclimit);
        }
        //pmesh->Rebalance();
 
        // some status update
        if ( myid == 0 and num_procs == 1 ) { 
        std::cout << "  successful! " << std::endl;
        }
        else if ( num_procs > 1 ) { 
            cout << "               MPI Rank " << myid << ":    Refined = " << setw(3) << num_refine << "   Total = " << pmesh->GetNE() << endl;
        }
        
        MPI_Barrier( MPI_COMM_WORLD );
    }

    pmesh->Finalize();

    num_bndr = pmesh->GetNBE();
    int num_wrong_bnd = pmesh->CheckBdrElementOrientation();
    if ( num_wrong_bnd > 0 ) {
        std::cout << "     WARNING: Boundary elements with wrong orientation found!" << std::endl;	
    }

    // Visualization of final refined mesh
    L2_FECollection attr_fec(0, dim);
    ParFiniteElementSpace attr_fespace(pmesh, &attr_fec);
    ParGridFunction attr(&attr_fespace);
    
    for ( int i = 0; i < pmesh->GetNE(); ++i ) {
        attr(i) = pmesh->GetAttribute(i);
        if ( attr(i) == 0 ) {
            std::cout << "     WARNING: Zero attribute found!" << std::endl;
        }
    }

    char vishost[] = "localhost";
    int  visport   = 19916;
    socketstream sol_sock(vishost, visport);

    sol_sock << "parallel " << num_procs << " " << myid << "\n";
    sol_sock.precision(8);
    sol_sock << "solution\n" << *pmesh << attr << "keys attciXXXXXXXXXXXXXXXXXXpppM" << flush; 

	// some timing data
	tic_toc.Stop();
	time_meshloading = tic_toc.RealTime();
	tic_toc.Clear();
	tic_toc.Start();

    // output mesh load time
    if (myid == 0) {
        std::cout << std::setprecision(2) << std::fixed;
        std::cout << "     Mesh load time [sec]: "  << time_meshloading << std::endl;
    }

	return 0;
}

```

**Comment by mlstowell** [comment]
Hi, @Maatinski,

Thanks for the test code. I was able to build and run it in parallel with no issues. On two processors it crashes my GLVis with a complaint about a missing "boundary" in the mesh stream. I'll try to take a closer look later and get back to you.

Best wishes,
Mark

**Comment by mlstowell** [comment]
Hi, @Maatinski,

I found a problem with your test code but I am still having trouble with my GLVis build so I cannot verify the output from the modified code. The issue is that `refine_any` may produce different values on different processors. Consequently different processors may exit the `while` loop prematurely. I would suggest adding an `MPI_Allreduce` to compute a globally valid flag such as `refine_any_global` and use this in the `while` and also when calling `pmesh->GeneralRefinement(refs, -1, nclimit);`.

This could perhaps explain the issue you are seeing but I wish my GLVis would display the results properly so that I could be sure (I probably have an outdated library somewhere; I don't think it's an issue with your test program).

Best wishes,
Mark

**Comment by Maatinski** [comment]
Mark, thank you so much for the helpful comments (and for crawling through my code)! 
I included the `MPI_Allreduce` as you suggested, and it seems to run more smoothly now (no crashes of my code itself, even on runs with many processes), but I am still having visualization issues with GLVIS, which either throws `MFEM abort: input stream is not a FiniteElementSpace!` or a `Verification failed: (ident == "boundary") is false:` error.

I am also guessing that there is something wrong with my boundary, because when I call `pmesh->Rebalance()` (after exiting the `while `loop) it somehow resets all attributes to the very initial version before NC refinement (even though the mesh elements themselves look good):

![image](https://user-images.githubusercontent.com/63864061/195442040-66c63795-9e3d-472a-bb01-3f16af17f11a.png)

Thanks a lot and warm regards,
Mathias

**Comment by mlstowell** [comment]
Hi, @Maatinski,

This is encouraging. I have not seen a GLVis crash when reading the finite element space but your  `Verification failed: (ident == "boundary") is false:` sounds like the same error I've seen.

I have not looked at the rebalance issue yet but now that I know we are seeing similar issues I'll dig into that.

Best wishes,
Mark

**Comment by mlstowell** [comment]
Hi, @Maatinski,

More progress. I think I tracked the somewhat unpredictable failures to a lack of `MPI_Finalize` in the test program. Without this execution would immediately halt when before all processors had completed writing their output.

I'm still seeing differences when rebalancing is used. It appears as though the attributes from the coarser mesh are still being used after refinement. I'm not sure why this is the case but I have an idea that I'm going to pursue.

Best wishes,
Mark

**Comment by mlstowell** [comment]
Hi, @Maatinski,

I cannot find how the `ParNCMesh` is moving attribute information between processes during rebalancing and judging by the GLVis output it isn't. My suspicion is that the `ParNCMesh` makes the assumption that the attribute number for the coarsest elements should be shared by their child elements. Of course this assumption is not valid in your meshes.

If I can confirm my suspicion I'll create an issue and hopefully we can find a suitable means to remove this assumption.

Best wishes,
Mark

**Comment by mlstowell** [comment]
Hi, @Maatinski,

Adding calls to the following code function we can see that the volumes are reverted by calls to rebalance.

This adds weight to my suspicion but the cause could still be elsewhere. One workaround you could employ would be to repaint the materials after each call to rebalance. This would certainly be suboptimal but it may help in the short term.

Best wishes,
Mark

```
void print_histogram()
{
   int myid = pmesh->GetMyRank();
   int max_attr = pmesh->attributes.Max();

   Vector h_loc(max_attr+1);
   h_loc = 0.0;

   const double *vert_1;
   const double *vert_2;
   const Element *cell;
   Array<int> vert_indices;

   for (int mc = 0; mc < pmesh->GetNE(); mc++)
   {
      cell = pmesh->GetElement(mc);
      cell->GetVertices(vert_indices);

      vert_1 = pmesh->GetVertex(vert_indices[0]);
      vert_2 = pmesh->GetVertex(vert_indices[6]);

      double dx = vert_2[0] - vert_1[0];
      double dy = vert_2[1] - vert_1[1];
      double dz = vert_2[2] - vert_1[2];

      int attr = pmesh->GetAttribute(mc);

      h_loc[attr] += dx * dy * dz * 1e6;
   }

   Vector h(max_attr+1);
   MPI_Reduce((double*)h_loc, (double*)h, max_attr+1,
              MPI_DOUBLE, MPI_SUM, 0, MPI_COMM_WORLD);

   if (myid == 0)
   {
      cout << "\nVolume by attribute ";
      for (int i=1; i<=max_attr; i++)
      {
         cout << " " << i << ":" << h[i];
      }
      cout << endl;
   }
}
```

**Comment by Maatinski** [comment]
Wow, thank you so much for the extensive help! 
(if you ever make it to the Boston area, I will owe you a beer)

Thanks for the debugging regarding the `ParNCMesh` not communicating children attributes: that's a little surprising though, given that the same behavior (children having different mesh attributes) should occur in the `shaper `miniapp as well, if I am not mistaken.
The "repainting" of the attributes after re-balancing works well, and doesn't really add too much overhead. I will start doing some more extensive testing on more realistic models, and also play around with removing the DoFs from the "background elements" as you suggested further above.

Big thanks,
Mathias

**Comment by mlstowell** [comment]
I wanted to ask @jakubcerveny if the suspicion I raised above is correct. What do you think Jakub, is it possible that user-defined element attributes would be overwritten during rebalancing? If so, do you think there might be a way to preserve user-defined attributes in such cases?

ping: @tzanio

**Comment by jakubcerveny** [comment]
Hi @mlstowell, I think you are right. It didn't occur to me before but we've actually never included the attributes in the rebalancing messages, because it was assumed they are the same as the attributes of the respective root elements. I'm not ever sure now if it is conceptually correct that the attributes of refined elements differ from their parents. If we're OK with that (i.e. there are no side effects elsewhere) we could add support for rebalancing with user-defined attributes. Another approach to store the material indicator data is using a piecewise-constant (L2) ParGridFunction where post-rebalancing Update() is supported.

**Comment by termi-official** [comment]
Sorry, I am a bit late to the party. I ran into the exact same problem in the past (#1891 and #1610 and references therein), because I do the exact same thing. Part of the problem is that the mesh stores a copy of the NC mesh leaves instead of a reference. However, attribute assignment is never synchronized between the NC mesh and the mesh (see https://github.com/mfem/mfem/blob/master/mesh/mesh.hpp#L1483 and https://github.com/mfem/mfem/blob/master/mesh/ncmesh.hpp#L474-L494 where the latter is never touched by `SetAttribute`). Note that this not only affects rebalancing, but also the mesh IO.

Quick question on the use case: If I would provide a full-hexahedral remesher for voxel data with subdomain support and good quality, would you use it or do you require the voxelized representation?

**Comment by Maatinski** [comment]
Hi @termi-official and @jakubcerveny ,
thanks for the clarifications! At the moment, I can bypass the issue by simply overwriting the attributes after the `Rebalance()` call. @termi-official not sure I fully understand what your voxel remesher does (or how it's different from what I am doing at the moment), but happy to have a look/give it a try. The main goal of this implementation is to reduce the memory load for these kind of meshes to be able to run this on our computers with limited memory.

The one thing I am currently trying to implement is getting rid of the "background" mesh elements (the ones that are left over from the refinement process, i.e., the blue cells in my very first post). It seems that in principle mfem can handle this kind of mesh, i.e., some parent elements having only a subset of the children (at the moment, I generate this kind of NC mesh from within Matlab and then read it as a compressed `.mesh.gz`, which works pretty well) but from what @mlstowell described earlier, it's currently not supported to remove these unneeded children elements. At the moment I am looking into dealing with this through boundary constraints; will let you know I am running into problems there..

In any case, I want to emphasize how helpful these NC meshes have been for us, especially in pushing the resolution of our E-field maps (which is needed to resolve the fine anatomical structures in our body models). Here is one of the more recent E-field maps; stitched together from two separate simulations (of course not the way to do it), hopefully we can run these soon in a single run.
![male_model_Efields](https://user-images.githubusercontent.com/63864061/202314513-5b57a64a-0a3c-40f3-8085-d60855f85c35.png)

**Comment by tzanio** [comment]
☝️ that's an amazing image @Maatinski !
